ckanext/datastore/backend/postgres.py

```diff
@@ -106,6 +120,63 @@ def __init__(self, error_dict: ErrorDict):
 _engines = {}
 
 
+# (canada fork only): parse constraint sql errors
+# TODO: upstream contrib!!
+class PartialFormat(Dict[Any, Any]):
+    def __missing__(self, key: str) -> str:
+        return "{" + key + "}"
```

Comment:
Why is this required? All the format values should be passed where this is used, right?

Reply:
in the core code I wrote here YES they are all passed. But in the error messages in Recombinant, we do not use the `{refKeys}`. So need to do a partial thing here to allow for those custom messages.

should I just make it so that you have to use all the replacement keys in the messages? or is this fine to do?

Reply:
it's just a little magical which makes the code harder to understand. Where is this used in recombinant?

Reply:
I put some comments in to explain what it does. And in Recombinant, the new code uses it from definitions from the Schemas.

Probably won't add this partial formatting to upstream though, just need it from specific custom error messages for our clients.